### PR TITLE
Add evidence to custom HTML reports

### PR DESCRIPTION
**Pull request by PeteBoucher**
This is my first contribution.

I added the checkbox and field as hinted in the issue report https://github.com/zaproxy/zaproxy/issues/6151

However I wasn't able to test my solution as I'm not sure how to install a locally built addOn. Also I couldn't find any tests to update. So any help testing this would be much appreciated!

Fix zaproxy/zaproxy#6151.

### CODE REVIEW

**Comment by psiinon** [comment]
Thanks for your contribution!
For building ZAP see https://github.com/zaproxy/zaproxy/wiki/Building and esp the gdoc it links to.
We should improve these docs :/

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

File: addOns/customreport/CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ### Changed
 - Update minimum ZAP version to 2.9.0.
 
+## 6 - 2020-09-02
+
+- Add Evidence to report (Issue 6151)
```
Comment (kingthorin) [inline addOns/customreport/CHANGELOG.md:16]:
```suggestion
### Changed
- Update minimum ZAP version to 2.9.0.
- Add Evidence to report (Issue 6151).
```

File: addOns/customreport/src/main/java/org/zaproxy/zap/extension/customreport/AlertDetailsPanel.java [hunk 2]
```diff
@@ -100,6 +102,12 @@ private void initialize() {
         gbc.gridy++;
         optionpanel.add(wascid, gbc);
 
+        // Include Evidence
+        evidence = new JCheckBox();
+        evidence.setText(Constant.messages.getString("customreport.alertdetails.evidence"));
```
Comment (kingthorin) [inline addOns/customreport/src/main/java/org/zaproxy/zap/extension/customreport/AlertDetailsPanel.java:107]: `customreport.alertdetails.evidence = Evidence` will need be to added to the [add-on's Messages.properties file](https://github.com/zaproxy/zap-extensions/blob/master/addOns/customreport/src/main/resources/org/zaproxy/zap/extension/customreport/resources/Messages.properties). (Line 34 I believe)

File: addOns/customreport/CHANGELOG.md (hunk 1 above)
Comment (kingthorin) [reply addOns/customreport/CHANGELOG.md:16]: Version and date are set during the release process.

File: addOns/customreport/src/main/java/org/zaproxy/zap/extension/customreport/AlertDetailsPanel.java (hunk 2 above)
Comment (PeteBoucher) [reply addOns/customreport/src/main/java/org/zaproxy/zap/extension/customreport/AlertDetailsPanel.java:107]: done

**Comment by PeteBoucher** [comment]
Yes please @kingthorin , I don't know where I picked that up, I may have signed too many commits on rebase.

**Comment by kingthorin** [comment]
Does that mean you do want me to fix the commits/history?

**Review by kingthorin — APPROVED** [review]

**Comment by PeteBoucher** [comment]
I have some problems when testing the CustomReports addOn. 

1. A warning pops up on Zap startup. A WARN messate is logged for a `ClassNotFoundException`
3. When opening "Customise HTML Report" dialog I get a Fatal error and the proper label is missing from the AlertDetails tab, not sure if I'm failing to compile the right project after making my changes to the properties file.
image
2. When generating a custom report the html report fails silently to write, no error is shown in the UI. A SAXParseException is output to console and the file `report_out.html-orig.xml` can be found where report_out.html was expected.

I'm a little lost here. I'm not familiar with Gradle so not sure if I'm testing this properly.

issue 1
```
23788 [ZAP-BootstrapGUI] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHttpPanelBrowserView
23792 [ZAP-BootstrapGUI] WARN org.zaproxy.zap.extension.browserView.ExtensionHttpPanelBrowserView  - Unable to use JavaFX:
java.lang.NoClassDefFoundError: javafx/embed/swing/JFXPanel
	at org.zaproxy.zap.extension.browserView.ExtensionHttpPanelBrowserView.isJavaFxAvailable(ExtensionHttpPanelBrowserView.java:95)
	at org.zaproxy.zap.extension.browserView.ExtensionHttpPanelBrowserView.hook(ExtensionHttpPanelBrowserView.java:75)
	at org.parosproxy.paros.extension.ExtensionLoader.hookAllExtension(ExtensionLoader.java:895)
	at org.parosproxy.paros.extension.ExtensionLoader.startLifeCycle(ExtensionLoader.java:794)
	at org.parosproxy.paros.control.AbstractControl.loadExtension(AbstractControl.java:58)
	at org.parosproxy.paros.control.Control.init(Control.java:137)
	at org.parosproxy.paros.control.Control.initSingletonWithView(Control.java:357)
	at org.zaproxy.zap.GuiBootstrap.initControlAndPostViewInit(GuiBootstrap.java:225)
	at org.zaproxy.zap.GuiBootstrap.access$100(GuiBootstrap.java:64)
	at org.zaproxy.zap.GuiBootstrap$2.run(GuiBootstrap.java:179)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException
	at org.zaproxy.zap.control.AddOnClassLoader.findClass(AddOnClassLoader.java:330)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 11 more
23798 [ZAP-BootstrapGUI] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Adds the Quick Start panel for scanning and exploring applications
```

issue 2
```
8207279 [AWT-EventQueue-0] ERROR org.zaproxy.zap.utils.I18N  - Failed to load a message:
java.util.MissingResourceException: Can't find resource for bundle java.util.PropertyResourceBundle, key customreport.alertdetails.evidence
	at java.base/java.util.ResourceBundle.getObject(ResourceBundle.java:564)
	at java.base/java.util.ResourceBundle.getString(ResourceBundle.java:521)
	at org.zaproxy.zap.utils.I18N.getStringImpl(I18N.java:181)
	at org.zaproxy.zap.utils.I18N.getString(I18N.java:148)
	at org.zaproxy.zap.extension.customreport.AlertDetailsPanel.initialize(AlertDetailsPanel.java:107)
	at org.zaproxy.zap.extension.customreport.AlertDetailsPanel.<init>(AlertDetailsPanel.java:48)
	at org.zaproxy.zap.extension.customreport.ExtensionCustomReport.getAlertDetailsTab(ExtensionCustomReport.java:138)
	at org.zaproxy.zap.extension.customreport.ExtensionCustomReport.access$400(ExtensionCustomReport.java:39)
	at org.zaproxy.zap.extension.customreport.ExtensionCustomReport$1.actionPerformed(ExtensionCustomReport.java:95)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.AbstractButton.doClick(AbstractButton.java:369)
	at java.desktop/com.apple.laf.ScreenMenuItem.actionPerformed(ScreenMenuItem.java:129)
	at java.desktop/java.awt.MenuItem.processActionEvent(MenuItem.java:690)
	at java.desktop/java.awt.MenuItem.processEvent(MenuItem.java:649)
	at java.desktop/java.awt.MenuComponent.dispatchEventImpl(MenuComponent.java:375)
```

issue 3
```
8413798 [AWT-EventQueue-0] ERROR org.zaproxy.zap.extension.customreport.ReportGenerator  - Open quote is expected for attribute "role" associated with an  element type  "a".
org.xml.sax.SAXParseException; lineNumber: 449; columnNumber: 55; Open quote is expected for attribute "role" associated with an  element type  "a".
	at java.xml/com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:261)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:339)
	at org.zaproxy.zap.extension.customreport.ReportGenerator.stringToHtml(ReportGenerator.java:111)
	at org.zaproxy.zap.extension.customreport.ReportLastScan.generateReport(ReportLastScan.java:401)
	at org.zaproxy.zap.extension.customreport.ExtensionCustomReport.generateReport(ExtensionCustomReport.java:124)
	at org.zaproxy.zap.extension.customreport.OptionDialog$2.actionPerformed(OptionDialog.java:87)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:270)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6654)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3345)
	at java.desktop/java.awt.Component.processEvent(Component.java:6419)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5029)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2773)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
8413866 [AWT-EventQueue-0] ERROR org.zaproxy.zap.extension.customreport.ReportGenerator  - /Users/pete/projects/owasp/report_out.html.temp (No such file or directory)
java.io.FileNotFoundException: /Users/pete/projects/owasp/report_out.html.temp (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:219)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:112)
	at java.base/java.io.FileReader.<init>(FileReader.java:60)
	at org.zaproxy.zap.extension.customreport.ReportGenerator.stringToHtml(ReportGenerator.java:155)
	at org.zaproxy.zap.extension.customreport.ReportLastScan.generateReport(ReportLastScan.java:401)
	at org.zaproxy.zap.extension.customreport.ExtensionCustomReport.generateReport(ExtensionCustomReport.java:124)
	at org.zaproxy.zap.extension.customreport.OptionDialog$2.actionPerformed(OptionDialog.java:87)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:270)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6654)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3345)
	at java.desktop/java.awt.Component.processEvent(Component.java:6419)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5029)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2773)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
8413890 [AWT-EventQueue-0] ERROR org.zaproxy.zap.utils.DesktopUtils  - Failed to mail or browse file:/Users/pete/projects/owasp/report_out.html. Error code: -43
java.io.IOException: Failed to mail or browse file:/Users/pete/projects/owasp/report_out.html. Error code: -43
	at java.desktop/sun.lwawt.macosx.CDesktopPeer.lsOpen(CDesktopPeer.java:169)
	at java.desktop/sun.lwawt.macosx.CDesktopPeer.browse(CDesktopPeer.java:74)
	at java.desktop/java.awt.Desktop.browse(Desktop.java:526)
	at org.zaproxy.zap.utils.DesktopUtils.openUrlInBrowser(DesktopUtils.java:42)
	at org.zaproxy.zap.extension.customreport.ReportLastScan.generateReport(ReportLastScan.java:414)
	at org.zaproxy.zap.extension.customreport.ExtensionCustomReport.generateReport(ExtensionCustomReport.java:124)
	at org.zaproxy.zap.extension.customreport.OptionDialog$2.actionPerformed(OptionDialog.java:87)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:270)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6654)
```

**Comment by kingthorin** [comment]
Number 1 is BrowserView not CustomReport.
Number 2 is CustomReport but is not up-to-date with this branch. 
```console
git fetch origin
git reset --hard origin/missing_evidence_6151
```
Number 3 I'm not sure about, likely specific to the ZAP session being used and some invalid content.

**Comment by thc202** [comment]
The number 3 might be caused by the missing `replaceEntity` (assuming the older changes) causing the XML to be malformed and the file to not be created.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thanks @PeteBoucher and @kingthorin!

**Review by psiinon — APPROVED** [review]

**Pull request merged**